torch/distributed/_composable/fsdp/fully_shard.py

```diff
@@ -205,3 +206,26 @@ def _get_fsdp_state(self) -> FSDPState:
         if (state := _get_module_fsdp_state(cast(nn.Module, self))) is None:
             raise AssertionError(f"No FSDP state found on {self}")
         return state
+
+    def _apply(self, *args: Any, **kwargs: Any) -> Any:
+        self.reshard()
+        ret = super()._apply(*args, **kwargs)  # type: ignore[misc]
```

Comment:
would  the recently landed `swap_tensor` keeps the original DTensor reference? it might not keep the `_local_tensor` reference though

Reply:
Good point! We should try to use the `swap_tensors` path for `_apply`, which swaps both the `DTensor` and the local tensor correctly. Let me hold off on this PR for a bit.